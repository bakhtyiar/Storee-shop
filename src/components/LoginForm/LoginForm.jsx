import React, {useContext} from 'react';
import {Button, Col, Form, Modal} from "react-bootstrap";
import {Link, useNavigate} from "react-router-dom";
import {RootContext} from "../../contexts/root-context/root-context";
import * as yup from "yup";
import {Formik} from "formik";
import {AuthModalContext} from "../../contexts/authModal-context/authModal-context";
import {BurgerMenuContext} from "../../contexts/burgerMenu-context/burgerMenu-context";
import {getCart} from "../../utils/server-api/cart/cart";
import {loginUser} from "../../utils/server-api/user/user";
import {routes} from "../../utils/constants";
import {initialState} from "../../contexts/root-context/initialState";
import FormTextField from "../formikElements/FormTextField";

//todo : remake login and register logic

const schema = yup.object().shape({
    username: yup.string()
        .min(6, 'Username is too short. Required minimum 6 symbols')
        .max(50, 'Username is too long. Required maximum 50 symbols')
        .required('Username is required'),
    password: yup.string()
        .min(6, 'Password is too short. Write at least 6 symbols')
        .required('Password is required'),
    forgetSession: yup.bool(),
});

const LoginForm = ({isHaveCloseButton = false}) => {
    const {authUserState: {onLogin}, cartState: {onSetCart}} = useContext(RootContext);
    const {onHide: hideBurgerMenu} = useContext(BurgerMenuContext);
    const {onSwitchType, onHide} = useContext(AuthModalContext);
    const navigate = useNavigate();

    const handleSubmit = async (values, actions) => {
        const {username, password} = values;
        const res = await loginUser(username, password);
        //tried to refactor & use .then.catch try{}catch(e){}, but cant get [[PromiseResult]] out from loginUser()
        if (res.message) { //message appears on 400 error
            actions.setFieldError('general', res.message);
            actions.setSubmitting(false);
            return ;
        }
        onLogin(res);
        const cart = await getCart(res.id);
        console.log('cart', cart);
        if (cart === undefined) {
            onSetCart(initialState.cartState);
        } else {
            onSetCart(cart);
        }
        onHide();
        hideBurgerMenu();
        navigate(`${routes.profile.path}`);
    };

    return (
        <Formik
            validationSchema={schema}
            validateOnBlur
            onSubmit={handleSubmit}
            initialValues={{
                username: '',
                password: '',
                forgetSession: false,
            }}
        >
            {({
                  handleSubmit,
                  handleChange,
                  handleBlur,
                  values,
                  touched,
                  errors,
              }) => (
                <Form noValidate onSubmit={handleSubmit}>
                    <Modal.Header closeButton={isHaveCloseButton}>
                        <Modal.Title id="contained-modal-title-vcenter">
                            Login
                        </Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <div className={'mb-3'}>
                            <FormTextField
                                as={Col}
                                controlId="formBasicUsername"
                                label='Username'
                                placeholder='Enter username'
                                name='username'
                                type='text'
                            />
                        </div>
                        <div className={'mb-3'}>
                            <FormTextField
                                as={Col}
                                controlId="formBasicPassword"
                                label='Password'
                                placeholder='Enter password'
                                name='password'
                                type='password'
                            />
                        </div>
                        <Form.Group className="mb-3" controlId="formBasicCheckbox">
                            <Form.Check
                                type="checkbox"
                                label="Do not remember session on this computer"
                                feedback={errors.forgetSession}
                                feedbackType="invalid"
                                name="terms"
                                onChange={handleChange}
                                onBlur={handleBlur}
                                isValid={touched.forgetSession && !errors.forgetSession}
                                isInvalid={touched.forgetSession && !!errors.forgetSession}
                            />
                        </Form.Group>
                        <h6 className='text-danger'>
                            {errors.general}
                        </h6>
                    </Modal.Body>
                    <Modal.Footer className='d-flex justify-content-between'>
                        <Link as={Button} onClick={onSwitchType}>Doesn't have an account?</Link>
                        <Button variant="primary" type="submit">
                            Login
                        </Button>
                    </Modal.Footer>
                </Form>
            )}
        </Formik>
    );
};

export default LoginForm;

