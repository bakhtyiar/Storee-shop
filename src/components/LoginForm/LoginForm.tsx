import React, {useContext} from 'react';
import {Button, Col, Form, Modal} from "react-bootstrap";
import {useNavigate} from "react-router-dom";
import {RootContext} from "../../contexts/root-context/root-context";
import * as yup from "yup";
import {Formik, FormikHelpers} from "formik";
import {AuthModalContext} from "../../contexts/authModal-context/authModal-context";
import {BurgerMenuContext} from "../../contexts/burgerMenu-context/burgerMenu-context";
import {getCart} from "../../utils/server-api/cart/cart";
import {loginUser} from "../../utils/server-api/user/user";
import {routes} from "../../utils/constants";
import {initialState} from "../../contexts/root-context/initialState";
import FormTextField from "../formikElements/FormTextField";

const schema = yup.object().shape({
    username: yup.string()
        .min(6, 'Username is too short. Required minimum 6 symbols')
        .max(50, 'Username is too long. Required maximum 50 symbols')
        .required('Username is required'),
    password: yup.string()
        .min(6, 'Password is too short. Write at least 6 symbols')
        .required('Password is required'),
    forgetSession: yup.bool(),
});

interface Props {
    isHaveCloseButton?: boolean,
    handleSubmit?: ((values: {
        general: string;
        username: string;
        password: string;
        forgetSession: boolean;
    }, formikHelpers: FormikHelpers<{
        general: string;
        username: string;
        password: string;
        forgetSession: boolean;
    }>) => void | Promise<any>) | null,
}

const LoginForm = ({isHaveCloseButton = false, handleSubmit = null}: Props) => {
    const {authUserState: {onLogin}, cartState: {onSetCart}} = useContext(RootContext);
    const {onHide: hideBurgerMenu} = useContext(BurgerMenuContext);
    const {onSwitchType, onHide} = useContext(AuthModalContext);
    const navigate = useNavigate();

    if (handleSubmit == null) {
        handleSubmit = async (values: any, actions: any) => {
            const {username, password} = values;
            let res;
            let cart;
            try {
                res = await loginUser(username, password);
                onLogin(res, values.forgetSession);
                cart = await getCart(res!.id);
                if (cart === undefined) {
                    onSetCart(initialState.cartState);
                } else {
                    onSetCart(cart);
                }
                onHide();
                hideBurgerMenu();
                navigate(`${routes.profile.path}`);
            } catch (e) {
                if (e instanceof Error) {
                    actions.setFieldError('general', e.message);
                    actions.setSubmitting(false);
                }
            }
        }
    }

    return (
        <Formik
            validationSchema={schema}
            validateOnBlur
            onSubmit={handleSubmit}
            initialValues={{
                general: '',
                username: '',
                password: '',
                forgetSession: false,
            }}
        >
            {({
                  handleSubmit,
                  handleChange,
                  handleBlur,
                  touched,
                  errors
              }) => (
                <Form noValidate onSubmit={handleSubmit} data-testid="login-form">

                    <Modal.Header closeButton={isHaveCloseButton}>

                        <Modal.Title id="contained-modal-title-vcenter">
                            Login
                        </Modal.Title>
                    </Modal.Header>

                    <Modal.Body>

                        <div className={'mb-3'}>

                            <FormTextField
                                as={Col}
                                controlId="formBasicUsername"
                                label='Username'
                                placeholder='Enter username'
                                name='username'
                                type='text'
                            />
                        </div>

                        <div className={'mb-3'}>

                            <FormTextField
                                as={Col}
                                controlId="formBasicPassword"
                                label='Password'
                                placeholder='Enter password'
                                name='password'
                                type='password'
                            />
                        </div>

                        <Form.Group className="mb-3" controlId="formBasicCheckbox">

                            <Form.Check
                                type="checkbox"
                                label="Do not remember session on this computer"
                                feedback={errors.forgetSession}
                                feedbackType="invalid"
                                name="terms"
                                onChange={handleChange}
                                onBlur={handleBlur}
                                isValid={touched.forgetSession && !errors.forgetSession}
                                isInvalid={touched.forgetSession && !!errors.forgetSession}
                            />
                        </Form.Group>

                        <h6 className='text-danger' data-testid="auth-error-feedback">
                            {errors.general}
                        </h6>
                    </Modal.Body>

                    <Modal.Footer className='d-flex justify-content-between'>

                        <Button variant="link" onClick={onSwitchType}>Doesn't have an account?</Button>

                        <Button variant="primary" type="submit" data-testid='submit-button'>
                            Login
                        </Button>
                    </Modal.Footer>
                </Form>
            )}
        </Formik>
    );
};

export default LoginForm;

