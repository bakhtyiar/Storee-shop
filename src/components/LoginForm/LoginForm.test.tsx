import {act, screen, waitFor} from "@testing-library/react";
import LoginForm from "./LoginForm";
import {renderWithRouter} from "../../../tests/integration-helpers/renderWithRouter";
import userEvent from "@testing-library/user-event";

describe("Basic tests", () => {
    test("Match snapshot", () => {
        renderWithRouter(<LoginForm/>);
        expect(screen.getByTestId("login-form")).toMatchSnapshot();
    })
})

describe("formik tests", () => {
    test("valid data", async () => {
        const handleSubmit = jest.fn(x => x);
        renderWithRouter(<LoginForm handleSubmit={handleSubmit}/>, "/someRandomRoute123");
        let loginField = screen.getByRole("textbox", {name: /name/i});
        await userEvent.type(loginField, "kminchelle");
        let passwordField = screen.getByText(/pass/i);
        await userEvent.type(passwordField, "0lelplR");
        let submitButton = screen.getByRole("button", {name: /login/i});
        await userEvent.click(submitButton);
        await waitFor(() => {
            expect(handleSubmit).toBeCalled();
        })
        await waitFor(() => {
            expect(handleSubmit.mock.results[0].value).toEqual(expect.objectContaining({
                username: "kminchelle",
                password: "0lelplR",
                forgetSession: false,
            }));
        })
    })
    test("no data", async () => {
        const handleSubmit = jest.fn(x => x);
        renderWithRouter(<LoginForm handleSubmit={handleSubmit}/>, "/someRandomRoute123");
        let submitButton = screen.getByRole("button", {name: /login/i});
        // eslint-disable-next-line testing-library/no-unnecessary-act
        await act(async () => {
            await userEvent.click(submitButton);
        })
        await waitFor(() => {
            expect(handleSubmit).not.toBeCalled();
        })
    })
    test("invalid data", async () => {
        const handleSubmit = jest.fn(x => x);
        renderWithRouter(<LoginForm handleSubmit={handleSubmit}/>, "/someRandomRoute123");
        let loginField = screen.getByRole("textbox", {name: /name/i});
        await userEvent.type(loginField, "k");
        let passwordField = screen.getByText(/pass/i);
        await userEvent.type(passwordField, "0l");
        let submitButton = screen.getByRole("button", {name: /login/i});
        await userEvent.click(submitButton);
        await waitFor(() => {
            expect(handleSubmit).not.toBeCalled();
        })
    })
})
