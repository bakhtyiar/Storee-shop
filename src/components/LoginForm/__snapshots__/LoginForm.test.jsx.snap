// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic tests Match snapshot 1`] = `
<form
  class=""
  data-testid="login-form"
  novalidate=""
>
  <div
    class="modal-header"
  >
    <div
      class="modal-title h4"
      id="contained-modal-title-vcenter"
    >
      Login
    </div>
  </div>
  <div
    class="modal-body"
  >
    <div
      class="mb-3"
    >
      <div
        class="col"
        data-testid="form-text-field"
      >
        <label
          class="form-label"
          for="formBasicUsername"
        >
          Username
        </label>
        <div
          class="input-group"
        >
          <input
            class="rounded form-control"
            id="formBasicUsername"
            name="username"
            placeholder="Enter username"
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="col"
        data-testid="form-text-field"
      >
        <label
          class="form-label"
          for="formBasicPassword"
        >
          Password
        </label>
        <div
          class="input-group"
        >
          <input
            class="rounded form-control"
            id="formBasicPassword"
            name="password"
            placeholder="Enter password"
            type="password"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="formBasicCheckbox"
          name="terms"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="formBasicCheckbox"
          title=""
        >
          Do not remember session on this computer
        </label>
      </div>
    </div>
    <h6
      class="text-danger"
    />
  </div>
  <div
    class="d-flex justify-content-between modal-footer"
  >
    <a
      as="[object Object]"
      href="/"
    >
      Doesn't have an account?
    </a>
    <button
      class="btn btn-primary"
      type="submit"
    >
      Login
    </button>
  </div>
</form>
`;
